lib/awful/widget/layoutlist.lua

```diff
@@ -0,0 +1,151 @@
+    return item.layout.name, selected and beautiful.bg_focus or nil, nil, item.icon
+end
+
+local function reload_cache(self)
+    self._private.cache = {}
+
+    for _, l in ipairs(alayout.layouts or {}) do
+        local icn_path, icon = beautiful["layout_" .. (l.name or "")]
+        if icn_path then
+            icon = surface.load(icn_path)
+        end
+
+        table.insert(self._private.cache, {
+            icon     = icon,
+            layout   = l,
+            screen   = self.screen,
+            callback = function() end,
+        })
+    end
+end
+
+local function update(self)
+    if not self._private.layout then return end
+    awcommon.list_update(
+        self._private.layout,
+        default_buttons,
+        wb_label,
+        self._private.data,
+        self._private.cache,
+        {
+            widget_template = self._private.widget_template
+        }
+    )
+end
+
+local layoutlist = {}
+
+--- The layoutlist layout.
+-- If no layout is specified, a `wibox.layout.fixed.vertical` will be created
+-- automatically.
+-- @property base_layout
+-- @param widget
+-- @see wibox.layout.fixed.vertical
+
+--- The delegate widget template.
+-- @property widget_template
+-- @param table
+
+--- The layoutlist screen.
+-- @property screen
+-- @param screen
+
+function layoutlist:set_base_layout(layout)
+    self._private.layout = layout
+
+    assert(layout.is_widget)
+
+    update(self)
+
+    self:emit_signal("widget::layout_changed")
+    self:emit_signal("widget::redraw_needed")
+end
+
+function layoutlist:set_widget_template(widget_template)
+    self._private.widget_template = widget_template
+
+    -- Remove the existing instances
+    self._private.data = {}
+
+    update(self)
+
+    self:emit_signal("widget::layout_changed")
+    self:emit_signal("widget::redraw_needed")
+end
+
+function layoutlist:layout(_, width, height)
+    if self._private.layout then
+        return { wibox.widget.base.place_widget_at(self._private.layout, 0, 0, width, height) }
+    end
+end
+
+function layoutlist:fit(context, width, height)
+    if not self._private.layout then
+        return 0, 0
+    end
+
+    return wibox.widget.base.fit_widget(self, context, self._private.layout, width, height)
+end
+
+--- Create an action list.
+--
+-- @tparam table args
+-- @tparam widget args.layout The widget layout (not to be confused with client
+--  layout).
+-- @tparam[opt] table args.widget_template A custom widget to be used for each action.
+-- @tparam screen args.screen A screen (mandatory)
+-- @treturn widget The action widget.
+-- @function naughty.widget.layoutlist
+
+local function new(_, args)
+    args = args or {}
+
+    local wdg = wibox.widget.base.make_widget(nil, nil, {
+        enable_properties = true,
+    })
+
+    gtable.crush(wdg, layoutlist, true)
+
+    reload_cache(wdg)
+
+    wdg._private.data = {}
+    gtable.crush(wdg, args)
+
+    tag.connect_signal("property::layout", function() reload_cache(wdg); update(wdg) end)
```

Comment:
Please wire up this widget to `test-leaks.lua` and then make the widget somehow GC'able.